# Homework2 Support Vector Machine
------


This is the README session for:
* Metric Number: `A0148008J`
* Email: `e0012680@u.nus.edu`

#### Common Programming snippet for writing the output into a file (given in the assignment)

In [2]:
def written_program_output (i, cost, kernel, gamma, degree):
    ### The output of your program should be written in a file as follows.
    #   for question 'i', write the output in 'problem-i.txt' file (e.g., 'problem-1a.txt')
    current_pwd = os.path.join(os.getcwd(),'homework/homework2')
    file_path = os.path.join(current_pwd, 'problem-'+i+'.txt')
    fo = open(file_path, 'w')

    # train your svm
    # (n.b., svmTrain, svmPredict are not previously defined;
    # you will have to supply code to implement them)
    svmModel, totalSV  = svmTrain(dataTrain, labelTrain, cost, kernel, gamma, degree)

    # test on the training data
    trainAccuracy = svmPredict(dataTrain, labelTrain, svmModel)

    # test on your test data
    testAccuracy = svmPredict(dataTest, labelTest, svmModel)

    # report your results in the file
    fo.write("Kernel: "+ str(kernel)+"\n")
    fo.write("Cost: "+ str(cost)+ "\n")
    fo.write("Number of Support Vectors: "+ str(totalSV)+"\n")
    fo.write("Train Accuracy: "+ str(trainAccuracy)+"\n")
    fo.write("Test Accuracy: " + str(testAccuracy)+"\n")

## Problem 1 - US Postal Service Zip Code dataset
------

In [6]:
import numpy as np
import os
from sklearn.svm import SVC

In [7]:
#Set up details
current_pwd = os.path.join(os.getcwd(),'homework/homework2')

train_data_path = os.path.join(current_pwd, 'hw2-1-train.txt')
test_data_path = os.path.join(current_pwd, 'hw2-1-test.txt')

In [8]:
# Load DataSet
setTrain = np.loadtxt(train_data_path)
setTest = np.loadtxt(test_data_path)
# split the data sets
labelTrain, dataTrain = setTrain[:,0], setTrain[:,1:3]
labelTest, dataTest = setTest[:,0], setTest[:,1:3]

##### Self-Defined Functions
-----

In [9]:
def svmTrain(dataTrain, labelTrain, cost, kernel, gamma, degree):
    model = SVC(C=cost, kernel=kernel, gamma=gamma, degree=degree)
    model.fit(dataTrain, labelTrain)
    return model, model.support_.shape[0]

In [10]:
def svmPredict(data, label, svmModel):
    predict = svmModel.predict(data)
    N = predict.shape[0]
    return 1 - np.sum(predict!=label)/(N*1.0)

In [11]:
def testAccuracy(dataTrain, labelTrain, dataTest, labelTest, cost, kernel, gamma, degree):
    model, totalSV = svmTrain(dataTrain, labelTrain, cost, kernel, gamma, degree)
    return svmPredict(dataTest, labelTest, model), model, totalSV

In [12]:
def printToVariousSubset(dataTrain, labelTrain, dataTest, labelTest, cost, kernel, gamma, degree):
    # {50, 100, 200, 800} of the subset of training data set
    # Linear kernel for model fitting
    for size in [50, 100, 200, 800]:
        dataSubsetTrain = dataTrain[0:size,:]
        labelSubsetTrain = labelTrain[0:size]
        print("The accuracy using the first "+ str(size)+" of the training set is: " + str(
            testAccuracy(dataSubsetTrain, labelSubsetTrain, dataTest, labelTest, cost, kernel, gamma, degree)[0]))

In [13]:
def printToVariousCostDegree(dataTrain, labelTrain, dataTest, labelTest, gamma):
    # Degree Q is 2 or 5
    # Cost has the following values: {0.0001, 0.001, 0.01, 1}
    # Polynomial kernel for model fitting
    kernel = 'poly'
    for cost in [0.0001, 0.001, 0.01, 1]:
        for degree in [2, 5]:
            out_err, model, totalSV = testAccuracy(dataTrain, labelTrain, dataTest, labelTest, cost, 
                                           kernel, gamma, degree)
            in_err = svmPredict(dataTrain, labelTrain, model)
            print("Cost: "+ str(cost)+"  Degree: "+ str(degree)
                  +"  in_error: "+str(1-in_err)+"  out_error: "+ str(1-out_err)
                  +"  total SV: "+str(totalSV))

In [14]:
def printToVariousCostRBF(dataTrain, labelTrain, dataTest, labelTest, gamma, degree):
    # Cost has the following values: C ∈ {0.01, 1, 10^2, 10^4, 10^6}
    # RBF kernel for model fitting
    # The rest parameter chosen to be the default values
    kernel = 'rbf'
    for cost in [0.01, 1, 100, 10000, 1000000]:
        out_err, model, totalSV = testAccuracy(dataTrain, labelTrain, dataTest, labelTest, cost, 
                                           kernel, gamma, degree)
        in_err = svmPredict(dataTrain, labelTrain, model)
        print("Cost: "+ str(cost) +"  in_error: "
              + str(1-in_err)+"  out_error: "+ str(1-out_err)
              +"  total SV: "+ str(totalSV))

#### Problem 1.a

In [15]:
# Train using whole training data set and write the result to file
written_program_output('1a', 1, 'linear', 'auto', 1)

In [16]:
printToVariousSubset(dataTrain, labelTrain, dataTest, labelTest, 1, 'linear', 'auto', 1)

The accuracy using the first 50 of the training set is: 0.971698113208
The accuracy using the first 100 of the training set is: 0.978773584906
The accuracy using the first 200 of the training set is: 0.981132075472
The accuracy using the first 800 of the training set is: 0.978773584906


#### Problem 1.b

In [17]:
printToVariousCostDegree(dataTrain, labelTrain, dataTest, labelTest, 'auto')

Cost: 0.0001  Degree: 2  in_error: 0.0224215246637  out_error: 0.0306603773585  total SV: 510
Cost: 0.0001  Degree: 5  in_error: 0.00640614990391  out_error: 0.0188679245283  total SV: 42
Cost: 0.001  Degree: 2  in_error: 0.0070467648943  out_error: 0.0188679245283  total SV: 152
Cost: 0.001  Degree: 5  in_error: 0.00512491992313  out_error: 0.0165094339623  total SV: 28
Cost: 0.01  Degree: 2  in_error: 0.00448430493274  out_error: 0.0188679245283  total SV: 54
Cost: 0.01  Degree: 5  in_error: 0.00448430493274  out_error: 0.0165094339623  total SV: 28
Cost: 1  Degree: 2  in_error: 0.00448430493274  out_error: 0.0188679245283  total SV: 25
Cost: 1  Degree: 5  in_error: 0.00448430493274  out_error: 0.0165094339623  total SV: 26


**_Statement_**
1. When C = 0.0001, Ein is higher at Q = 5.
2. When C = 0.001, the number of support vectors is lower at Q = 5.
3. When C = 0.01, Ein is higher at Q = 5.
4. When C = 1, Eout is lower at Q = 5.
5. None of the above

**_Answer_**
1. False, Ein is lower at Q=5 when C=0.0001
2. True, when C=0.001, number of support vector is 152 for Q=2 which is higher than 28 for Q=5
3. False, Ein is the same for both, which is 0.00448430493274
4. True, Eout is about 0.002 lower for Q=5

#### Problem 1.c

In [18]:
printToVariousCostRBF(dataTrain, labelTrain, dataTest, labelTest, 'auto', 2)

Cost: 0.01  in_error: 0.00384368994234  out_error: 0.0212264150943  total SV: 347
Cost: 1  in_error: 0.00448430493274  out_error: 0.0212264150943  total SV: 31
Cost: 100  in_error: 0.00320307495195  out_error: 0.0188679245283  total SV: 20
Cost: 10000  in_error: 0.00256245996156  out_error: 0.0188679245283  total SV: 17
Cost: 1000000  in_error: 0.00128122998078  out_error: 0.0212264150943  total SV: 18


**_Question_**  
  
Which value of C ∈ {0.01, 1, 100, 104, 106} results in the lowest Ein? The lowest Eout?

**_Anwser_**
1. For Ein, when C=10^6, it attends the lowest value, which is about 0.00013
2. For Eout, when C=100 and C=10^4, it attends the lowest value, which is about 0.0189

## Statement of Individual Work

Please initial (between the square brackets) one of the following statements.

[Dong Shaocong] I, <*A0148008J*>, certify that I have followed the CS 3244 Machine Learning class guidelines for homework assignments.  In particular, I expressly vow that I have followed the Facebook rule in discussing with others in doing the assignment and did not take notes (digital or printed) from the discussions.  
